planner/core/preprocess.go

```diff
@@ -1312,3 +1342,49 @@ func (p *preprocessor) checkFuncCastExpr(node *ast.FuncCastExpr) {
 		}
 	}
 }
+
+func (p *preprocessor) handleAsOf(node *ast.AsOfClause) {
+	dom := domain.GetDomain(p.ctx)
+
+	tso := uint64(0)
+	if node != nil {
+		tsRes, err := evalAstExpr(p.ctx, node.TsExpr)
+		if err != nil {
+			p.err = err
+			return
+		}
+
+		toTypeTimestamp := types.NewFieldType(mysql.TypeTimestamp)
+		// We need at least the millionsecond here, so set fsp to 3.
+		toTypeTimestamp.Decimal = 3
+		ts, err := tsRes.ConvertTo(p.ctx.GetSessionVars().StmtCtx, toTypeTimestamp)
+		if err != nil {
+			p.err = err
+			return
+		}
+
+		tsTime, err := ts.GetMysqlTime().GoTime(p.ctx.GetSessionVars().TimeZone)
+		if err != nil {
+			p.err = err
+			return
+		}
```

Comment:
I think we can reuse the `calculateTsExpr` function defined in [planbuilder.go#L2345]( https://github.com/pingcap/tidb/blob/master/planner/core/planbuilder.go#L2345) to do the same thing. A little modification to that function signature is needed though.

Reply:
Yeah, but that function must be extracted somewhere.